From 2cb604ec596f171b621e6076713dbf7e4bc3a888 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:40:14 -0400
Subject: [ethernet] i40e: Give link more time after setting flow control

Message-id: <1409658042-22689-254-git-send-email-sassmann@redhat.com>
Patchwork-id: 90455
O-Subject: [RHEL7.1 PATCH 253/281] i40e: Give link more time after setting flow control
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Catherine Sullivan <catherine.sullivan@intel.com>

Give link a little more time to come back up after setting flow control
before resetting. In the new NVMs it is taking longer for link to come back.
This causes the driver to attempt to reset the link, which then errors
because the firmware was already in the middle of a reset. Also, initialize
err to 0.

Change-ID: I1cc987a944e389d8909c262da5796f50722b4d6b
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Jim Young <jmyoungx@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 7d62dac6312efa7824eb59d8161aee8fef1c166c)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
index 055f897..3884c6c 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ethtool.c
@@ -630,7 +630,7 @@ static int i40e_set_pauseparam(struct net_device *netdev,
  bool link_up = hw_link_info->link_info & I40E_AQ_LINK_UP;
  i40e_status status;
  u8 aq_failures;
- int err;
+ int err = 0;
 
  if (vsi != pf->vsi[pf->lan_vsi])
   return -EOPNOTSUPP;
@@ -683,8 +683,12 @@ static int i40e_set_pauseparam(struct net_device *netdev,
   err = -EAGAIN;
  }
 
- if (!test_bit(__I40E_DOWN, &pf->state))
-  return i40e_nway_reset(netdev);
+ if (!test_bit(__I40E_DOWN, &pf->state)) {
+  /* Give it a little more time to try to come back */
+  msleep(75);
+  if (!test_bit(__I40E_DOWN, &pf->state))
+   return i40e_nway_reset(netdev);
+ }
 
  return err;
 }
-- 
1.7.1